# Awesome Array Driver

----

In [1]:
import os
os.chdir('..')

import aad

help(aad)

Help on package aad:

NAME
    aad

PACKAGE CONTENTS
    lld

CLASSES
    builtins.object
        AwesomeArrayDriver
    
    class AwesomeArrayDriver(builtins.object)
     |  AwesomeArrayDriver(pin=22336)
     |  
     |  Awesome Array Driver
     |  
     |  ...
     |  Attributes
     |  ----------
     |  _lld: lld.LLDriver
     |          The low-level driver used for the µc
     |  
     |  Methods defined here:
     |  
     |  __init__(self, pin=22336)
     |      Creates the driver.
     |      
     |      Details:
     |              It will search for the µc using the PID value 'DEFAULT_PID' or the one provided in argument.
     |              Takes the first found if many have the same PID.
     |              RAISE if not found.
     |      
     |      Arguments:
     |              pid: optional, the pid to search for.
     |  
     |  configure_sr(self, col: int, row: int, bar: bool, set: bool)
     |      Configure the shift registers for the given memristor location 

---

In [2]:
print("---- List of available ports ----")
aad.print_ports()

---- List of available ports ----
COM3 - Intel(R) Active Management Technology - SOL (COM3) | PID:  None
COM1 - Communications Port (COM1) | PID:  None
COM6 - USB Serial Device (COM6) | PID:  22336
COM5 - STMicroelectronics STLink Virtual COM Port (COM5) | PID:  14155


The communication with the µc is done via `USB Serial Device` at `COM6`. `COM5` is used only for power and its programming (i.e. not interesting for us).  
The driver will search by default for a device with `PID` 22336.

In [3]:
driver = aad.AwesomeArrayDriver()

print("The driver selected the port:", driver._lld.ser.port)

The driver selected the port: COM6


If no device with such `PID` was found, an exception would be raised.  
You can then check the connection, or use `aad.print_ports` to find the correct `PID` and give it as an argument of the constructor:  
```
driver = aad.AwesomeArrayDriver(correct_pid)
```

---  
(For now, the driver only configures the shift registers)  
A memristor location is given by three parameters:
* `col`: integer : Address of the column in the array
* `row`: integer : Address of the row in the array
* `bar`: boolean : Whether to select the memristor Rb (if True) or R (otherwise)  

Then, you can set the individual state of a memristor using the functions `set` or `reset`.  
You can also form the memristor first using the function `form`.  

For a more complete description, see the `help` output above.

----

In [4]:
# Sanity check of the shift registers for one particular configuration
sanity = driver.test_sr_sanity(col=0, row=0, bar=False, set=True)
for sr_id in aad.SR_list:
	for bit_id in range(64):
		# Usually, the condition would be "not ..." but for this test, the output is random and most registers will appear as not sane
		if sanity[sr_id][bit_id]:
			print(f"{aad.get_sr_name(sr_id)}'s bit #{bit_id} is sane")

WLE's bit #0 is sane
SL's bit #0 is sane
BLB's bit #0 is sane


In [5]:
driver.set(col=0, row=0, bar=False)
driver.reset(col=0, row=0, bar=True)